File: src/app/services/room.service.ts
import { Answer, Game, GamePhase, Vote } from '../models/game.model';
import { User } from '../models/user.model';

type RoomMap = Record<string, Game>;

const ROOM_STORAGE_KEY = 'rooms';

// Load preexisting rooms from local storage
function loadRooms(): RoomMap {
    const roomsJson = localStorage.getItem(ROOM_STORAGE_KEY);

    if (!roomsJson) {
        return {};
    }

    try {
        const parsedRooms = JSON.parse(roomsJson) as RoomMap;
        return parsedRooms && typeof parsedRooms === 'object' ? parsedRooms : {};
    } catch {
        return {};
    }
}

// Save all rooms into local storage
function saveRooms(rooms: RoomMap): void {
    localStorage.setItem(ROOM_STORAGE_KEY, JSON.stringify(rooms));
}

// Add a room into the room map
function saveRoom(game: Game): void {
    const rooms = loadRooms();
    rooms[game.roomName] = game;
    saveRooms(rooms);
}

// Load a room via code
function loadRoom(roomCode: string): Game | null {
    const rooms = loadRooms();
    return rooms[roomCode] ?? null;
}

// Create a new room
function createRoom(roomCode: string): Game | null {
    const game: Game = { roomName: roomCode, phase: 'waiting', answers: [], players: [] };
    saveRoom(game);
    return game;
}

// Given a code, either create or join a preexisting room
function createOrJoinRoom(roomCode: string, user: User): Game | null {
    console.log(' createing or joining room', roomCode);
    const existingRoom = loadRoom(roomCode);

    if (existingRoom) {
        return joinRoom(existingRoom, user);
    }

    const newRoom = createRoom(roomCode);

    if (!newRoom) {
        return null;
    }

    return joinRoom(newRoom, user);
}

// Add a user to a game's player list
function joinRoom(game: Game, user: User): Game {
    const isAlreadyInRoom = game.players.some(
        (player) => player.playerId === user.playerId,
    );

    if (!isAlreadyInRoom) {
        game.players.push(user);
        saveRoom(game);
    }

    return game;
}

// Remove a user from a game's player list, if the game is empty delete it
function leaveRoom(roomCode: string, playerId: string): Game | null {
    const room = loadRoom(roomCode);

    if (!room) {
        return null;
    }

    room.players = room.players.filter((player) => player.playerId !== playerId);

    if (room.players.length === 0) {
        closeRoom(roomCode);
        return null;
    }

    saveRoom(room);
    return room;
}

// Update the screen every player in a room should see next
function updatePhase(roomCode: string, phase: GamePhase): Game | null {
    const room = loadRoom(roomCode);

    if (!room) {
        return null;
    }

    room.phase = phase;
    saveRoom(room);
    return room;
}

// Save one answer for each player and move the room to the next phase.
function submitAnswers(roomCode: string, answers: Answer[]): Game | null {
    const room = loadRoom(roomCode);

    if (!room) {
        return null;
    }

    room.answers = answers;
    room.phase = 'judging';
    saveRoom(room);
    return room;
}

// Record one player's vote for another player's submitted responses.
function voteForAnswer(
    roomCode: string,
    answerPlayerId: string,
    voterId: string,
    vote: Vote,
): Game | null {
    const room = loadRoom(roomCode);

    if (!room) {
        return null;
    }

    const answer = room.answers.find((item) => item.playerId === answerPlayerId);

    if (!answer) {
        return null;
    }

    answer.votes ??= {};
    const previousVote = answer.votes[voterId];

    if (previousVote === vote) {
        return room;
    }

    if (previousVote === 'up') answer.score -= 1;
    if (previousVote === 'down') answer.score += 1;

    answer.votes[voterId] = vote;
    answer.score += vote === 'up' ? 1 : -1;
    saveRoom(room);
    return room;
}

// Close a room
function closeRoom(roomCode: string): void {
    const rooms = loadRooms();
    delete rooms[roomCode];
    saveRooms(rooms);
}

// Get total active games
function totalGames(): number {
    const rooms = loadRooms();
    return Object.keys(rooms).length;
}

// Get number of players in all games
function totalPlayers(): number {
    const rooms = loadRooms();
    let playerCount = 0;

    for (const room of Object.values(rooms)) {
        playerCount += room.players.length;
    }

    return playerCount;
}

// Delete rooms (for testing)
function clearRooms(): void {
    localStorage.removeItem(ROOM_STORAGE_KEY);
}

export const RoomService = {
    createOrJoinRoom,
    joinRoom,
    loadRoom,
    leaveRoom,
    updatePhase,
    submitAnswers,
    voteForAnswer,
    totalGames,
    totalPlayers,
    clearRooms
};

File: backend/room_service.py
from models import Answer, Game, GamePhase, Player, Vote

CURRENT_ROOMS: dict[str, Game] = {}


def load_room(room_code: str) -> Game | None:
	return CURRENT_ROOMS.get(room_code)


def create_room(room_code: str) -> Game:
	room = Game(roomName=room_code)
	CURRENT_ROOMS[room_code] = room
	return room


def create_or_join_room(room_code: str, player: Player) -> Game:
    existing_room = load_room(room_code)

    if existing_room is not None:
        return join_room(existing_room, player)

    new_room = create_room(room_code)
    return join_room(new_room, player)


def join_room(room: Game, player: Player) -> Game:
    already_in_room = any(
        existing_player.playerId == player.playerId
        for existing_player in room.players
    )

    if not already_in_room:
        player.currentRoom = room.roomName
        room.players.append(player)

    return room


def leave_room(room_code: str, player_id: str) -> Game | None:
	room = load_room(room_code)

	if room is None:
		return None

	room.players = [player for player in room.players if str(player.playerId) != player_id]

	if not room.players:
		close_room(room_code)
		return None

	return room


def update_phase(room_code: str, phase: GamePhase) -> Game | None:
	room = load_room(room_code)

	if room is not None:
		room.phase = phase

	return room


def submit_answers(room_code: str, answers: list[Answer]) -> Game | None:
	room = load_room(room_code)

	if room is not None:
		room.answers = answers
		room.phase = GamePhase.JUDGING

	return room


def vote_for_answer(
	room_code: str,
	answer_player_id: str,
	voter_id: str,
	vote: Vote,
) -> Game | None:
	room = load_room(room_code)

	if room is None:
		return None

	answer = next(
		(item for item in room.answers if str(item.playerId) == answer_player_id),
		None,
	)

	if answer is None:
		return None

	previous_vote = answer.votes.get(voter_id)

	if previous_vote == vote:
		return room

	if previous_vote == Vote.UP:
		answer.score -= 1
	elif previous_vote == Vote.DOWN:
		answer.score += 1

	answer.votes[voter_id] = vote
	answer.score += 1 if vote == Vote.UP else -1
	return room


def close_room(room_code: str) -> None:
	CURRENT_ROOMS.pop(room_code, None)


def total_games() -> int:
	return len(CURRENT_ROOMS)


def total_players() -> int:
	return sum(len(room.players) for room in CURRENT_ROOMS.values())


def clear_rooms() -> None:
	CURRENT_ROOMS.clear()

